File: Backtracking/46.permutations.py
#
# @lc app=leetcode id=46 lang=python3
#
# [46] Permutations
#

# @lc code=start
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        def backtracking(path, used):
            if len(path) == len(nums):
                res.append(path[:])
                return

            for i in range(len(nums)):
                if used[i] == 0:
                    used[i] = 1
                    path.append(nums[i])
                    backtracking(path, used)
                    used[i] = 0
                    path.pop()

        initUsed = [0] * len(nums)
        res = []
        backtracking([], initUsed)
        return res
# @lc code=end

File: Backtracking/46.permutations.java
/*
 * @lc app=leetcode id=46 lang=java
 *
 * [46] Permutations
 */

// @lc code=start

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

class Solution {
    List<List<Integer>> res = new ArrayList<>();
    LinkedList<Integer> path = new LinkedList<>();

    void helper(int[] nums, boolean isUsedNums[]) {
        if (path.size() == nums.length) {
            res.add(new ArrayList<>(path));
            return;
        }
        for (int i = 0; i < nums.length; i++) {
            if (!isUsedNums[i]) {
                isUsedNums[i] = true;
                path.add(nums[i]);
                helper(nums, isUsedNums);
                isUsedNums[i] = false;
                path.removeLast();
            }
        }
    }

    public List<List<Integer>> permute(int[] nums) {
        boolean[] isUsedNums = new boolean[nums.length];
        helper(nums, isUsedNums);
        return res;
    }
}
// @lc code=end
